pkg/common/fflag/fflag.go

```diff
@@ -0,0 +1,126 @@
+// The fflag package implements a basic singleton pattern for the purpose of
+// providing SPIRE with a system-wide feature flagging facility. Feature flags
+// can be easily added here, in a single central location, and be consumed
+// throughout the codebase.
+package fflag
+
+import (
+	"errors"
+	"fmt"
+	"sync"
+)
+
+// Flag represents a feature flag and its configuration name
+type Flag string
+
+// To add a feature flag, declare it in `common.go`, `agent.go`, or
+// `server.go` depending on where it should be considered valid. Then,
+// add it to the relevant package-level map, setting the appropriate
+// default value. Common feature flags are added to the map in `common.go`,
+// while agent or server specific flags are added to the map in the `afflag` or
+// `sfflag` package, located in the agent or server codebase, respectively.
+// Flags should always be opt-in, and default to false, with the only exception
+// being flags that are in the process of being deprecated.
+var (
+	singleton = struct {
+		flags  map[Flag]bool
+		loaded *sync.Once
+		mtx    *sync.RWMutex
+	}{
+		flags:  commonFlagMap,
```

Comment:
Yes, I thought this would be ok because it is unexported.

I had some code that copied the commonFlagMap into this flags field as a new map, but it got a little messy and cumbersome and since commonFlagMap is not exported I thought why not just re-use it as flags field and mutate it... this was previously statically defined here (and mutated) but then the elements moved into a different map in `common.go` following the server/agent split

What is the advantage of copying the commonFlagMap elements into a new singleton.flags map?

Reply:
To be clear, I'm not suggesting we copy it into singleton.flags here. I suggest commonFlagMap be passed or explicitly handled by `load()` to build the resulting flag set and `singleton.flags` be set from that returned value.

Also, I don't think anything is broken with the way it is currently done, but aliasing and unintentionally modifying this map that would otherwise likely be a constant if the language allowed for it, is just confusing and could potentially cause bugs in future refactors.

Reply:
> To be clear, I'm not suggesting we copy it into singleton.flags here. I suggest commonFlagMap be passed or explicitly handled by load() to build the resulting flag set and singleton.flags be set from that returned value.

Gotcha .. this was/is difficult to do due to the current structure of the validate/load code and associated locking (which is why I opted for this shortcut), but with the help of a shakeup in Load (which will come by way of other CR comments you'e left), I think that is the right direction. Thank you!

> Also, I don't think anything is broken with the way it is currently done, but aliasing and unintentionally modifying this map that would otherwise likely be a constant if the language allowed for it, is just confusing and could potentially cause bugs in future refactors.

yes ,. it would have been a constant, and I would not have been allowed to do such trickeries! 😆